### PR TITLE
Add ability to override `end2end` mode

### PR DESCRIPTION
**Pull request by Y-T-G**
```bash
yolo export model=yolo26n.pt end2end=False
yolo export model=yolo26n.pt format=coreml end2end=False nms  # Export with CoreML NMS
yolo val model=yolo26n.pt data=coco128.yaml end2end=False  # Validate with one2many head
yolo predict model=yolo26n.pt data=coco128.yaml end2end=False  # Predict with one2many head
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds an `end2end` switch to easily toggle YOLO26/YOLOv10 between NMS-free (one-to-one) and traditional NMS (one-to-many) pipelines across predict/val/export 🚀

### 📊 Key Changes
- 🧠 **Documented YOLO26 dual-head architecture** in the YOLO26 model docs, explaining:
  - **One-to-One head (default)**: NMS-free, capped detections per image, simpler deployment
  - **One-to-Many head**: classic dense outputs that require NMS, often slightly higher accuracy
- 🎛️ **Introduced `end2end` argument** (optional) for **prediction, validation, and export**, enabling users to override a model’s head behavior at runtime.
- ⚙️ **Config support added**:
  - `end2end` added to `default.yaml` and recognized in config arg keys.
  - `export-args.md`, `predict-args.md`, and `validation-args.md` updated to document `end2end`.
- 🔧 **Core runtime wiring**:
  - `Exporter`, `Predictor`, and `Validator` now apply `args.end2end` to the model when supported.
  - Added proper **getter/setter support** for `end2end` in the detection head and tasks wrapper to allow safe overriding.
- 🧪 **Expanded test coverage**:
  - CLI and export tests now run with **both `end2end=True` and `end2end=False`**.
  - Export test matrices avoid invalid combos like **`end2end=True` with `nms=True`** (since end-to-end mode implies NMS-free). 🧯

### 🎯 Purpose & Impact
- 🧩 **Simplifies deployment choices**: users can pick **fast NMS-free inference** or **traditional NMS-based workflows** without changing models or code structure.
- 🔄 **Better compatibility with existing pipelines**: setting `end2end=False` enables classic postprocessing and unlocks NMS-related knobs like `iou` and `agnostic_nms`. 🛠️
- 📦 **More predictable exports**: ONNX/TorchScript/OpenVINO/CoreML/TFLite/MNN exports can be aligned with the target runtime’s expected postprocessing behavior.
- ✅ **Improved reliability**: broader automated testing reduces regressions when switching between head modes, especially for YOLO26 (recommended) and YOLOv10. 📈

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help kick things off—an Ultralytics engineer will assist soon. 🙌

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean. The end2end override is applied consistently across predict/val/export and the head/task setters preserve backward compatibility. No issues found.

File: ultralytics/engine/predictor.py [hunk 1]
```diff
@@ -387,6 +387,8 @@ def setup_model(self, model, verbose: bool = True):
             model (str | Path | torch.nn.Module, optional): Model to load or use.
             verbose (bool): Whether to print verbose output.
         """
+        if hasattr(model, "end2end") and self.args.end2end is not None:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/predictor.py:390]: 💡 **MEDIUM**: This override is applied before `AutoBackend` is constructed. When `model` is a string/path (the common CLI case), `hasattr(model, "end2end")` is false and the override never reaches the actual model. Consider moving the assignment to after `self.model = AutoBackend(...)` and applying it to `self.model` instead.

File: ultralytics/engine/validator.py [hunk 2]
```diff
@@ -156,6 +156,8 @@ def __call__(self, trainer=None, model=None):
             if str(self.args.model).endswith(".yaml") and model is None:
                 LOGGER.warning("validating an untrained model YAML will result in 0 mAP.")
             callbacks.add_integration_callbacks(self)
+            if hasattr(model, "end2end") and self.args.end2end is not None:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/validator.py:159]: 💡 **MEDIUM**: Same issue as predictor: if `model` is a string/path, this block won't run and the override won't be applied to the `AutoBackend` instance. The override should likely be set on `model` after `AutoBackend(...)` is created.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good feature addition and minimal changes. The main concern is that the `end2end` override is set before `AutoBackend` construction in predict/val flows, so it won’t take effect for typical string/path inputs. Consider moving the assignment to after model instantiation. Otherwise looks clean.

💬 Posted 2 inline comments

File: ultralytics/engine/predictor.py (hunk 1 above)
Comment (Y-T-G) [reply ultralytics/engine/predictor.py:390]: Too messy if we want to handle this.

There's `end2end` attribute in 3 places: inside `AutoBackend`, `DetectionModel` and the `Detect` head. We would need to add an extra argument to `AutoBackend` to support different initialization for `end2end` override. Because the `end2end` modification needs to be applied before `fuse()` which happens during initialization otherwise the required layers are removed,

**Comment by Laughing-q** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-01-21 10:06:42 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/21205367544): `ci.yml` run 17000
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/21205368232): `docker.yml` run 1715

**Comment by Laughing-q** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-01-21 11:36:39 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/21208074554): `ci.yml` run 17008
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/21208075081): `docker.yml` run 1717

**Comment by Laughing-q** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-01-22 04:37:49 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/21236230957): `ci.yml` run 17046
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/21236231221): `docker.yml` run 1721

**Review by Laughing-q — APPROVED** [review]

**Review by Laughing-q — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this PR merged — huge thanks @Y-T-G (and @Laughing-q for the contributions) for a thoughtful, high-impact improvement to Ultralytics!  

> “Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away.” — Antoine de Saint-Exupéry  

This change embodies that idea: a clean `end2end` switch that makes it effortless to choose between YOLO26’s NMS-free one-to-one pipeline and the traditional one-to-many + NMS workflow across predict/val/export—without forcing users to restructure code or swap models. The documentation upgrades, safe runtime wiring, and expanded tests (including avoiding invalid `end2end=True` + `nms=True` combos) make this both elegant and robust.  

Really appreciate the care you put into usability, clarity, and reliability—this will make deploying YOLO26 via the [Ultralytics Platform](https://platform.ultralytics.com) and beyond even smoother.
